File: core/src/main/java/de/qabel/core/repository/sqlite/migration/AbstractMigration.java
package de.qabel.core.repository.sqlite.migration;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public abstract class AbstractMigration {
    private Connection connection;

    public AbstractMigration(Connection connection) {
        this.connection = connection;
    }

    public abstract long getVersion();

    protected void execute(String sql, Object... parameters) throws SQLException {
        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            for (int i = 1; i < parameters.length; i++) {
                statement.setObject(i, parameters[i]);
            }
            statement.execute();
        }
    }

    public abstract void up() throws SQLException;
    public abstract void down() throws SQLException;
}

File: core/src/main/java/de/qabel/core/repository/sqlite/migration/AbstractMigration.kt
package de.qabel.core.repository.sqlite.migration

import de.qabel.core.repository.sqlite.use
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.SQLException

abstract class AbstractMigration(private val connection: Connection) {

    abstract val version: Long

    @Throws(SQLException::class)
    protected fun execute(sql: String, vararg parameters: Any) {
        connection.prepareStatement(sql).use { statement ->
            for (i in 1..parameters.size - 1) {
                statement.setObject(i, parameters[i])
            }
            statement.execute()
        }
    }

    @Throws(SQLException::class)
    abstract fun up()

    @Throws(SQLException::class)
    abstract fun down()
}
